### ISSUE TITLE
No semantic token information for macro parameters

### ISSUE THREAD

**Issue by akbyrd** [microsoft/vscode-cpptools#10547]
### Environment

OS and Version: Windows 10 22H2 19045
VS Code Version: 1.75.1
C/C++ Extension Version: 1.14.3

### Bug Summary and Steps to Reproduce

Bug Summary:

The built-in textmate grammar attempts to detect macro parameters. However, it's not particularly successful. This is a place where semantic token information from the extension would help.

In this image examples 1 and 3 are incorrect.
In example 1 `param` is `meta.preprocessor.macro.cpp`.
In example 2 `#param` is correctly recognized as `variable.other.macro.argument.cpp`. I'm unsure if it should contain the `#` though.
In example 3 `#bar` is incorrectly marked as `variable.other.macro.argument.cpp`.
![image](https://user-images.githubusercontent.com/484784/219936597-2a5dc0a5-22e1-438d-bd1d-3b091ba3f930.png)

I think the desired outcome here is to mark `param` as semantic `variable.macro` and map that to syntax `variable.other.macro.argument.cpp`, or at least something roughly equivalent. And also to not mark `bar` as a variable.

### Configuration and Logs

```shell
N/A
```

**Comment by sean-mcmanus** [comment]
@akbyrd Is there a way to get the `#bar` to compile? The semantic colorization feature is meant for cases where proper handling require full parsing and semantic information and not to deal with issues that could be properly handled by the lexical colorization...not sure which this issue applies to yet.

**Comment by akbyrd** [comment]
I don't believe so. I believe the token after it is required to be a macro parameter. I can report the issue with `#bar` to https://github.com/jeff-hykin/better-cpp-syntax.

Should the first case be handled syntactically or semantically?

**Comment by sean-mcmanus** [comment]
@akbyrd What do you mean by "first case"? I'm guessing you're referencing `#define FOO1(param) param`. Our IntelliSense engine won't give semantic tokens for macros since they're not compiled on their own, they can only be handled lexically.

**Comment by leidegre** [comment]
I'll chime in here with my two cents.

Example,

![image](https://github.com/microsoft/vscode-cpptools/assets/63085/6a6fb9bc-9f7c-46ab-a588-3f9395991df4)

The TextMate scope variable.parameter.preprocessor is applied to the parameter but it doesn't stick with `expr` inside the macro definition `expr` has a different scope meta.preprocessor.macro.

Ideally, I'd like this to to work consistently.

![image](https://github.com/microsoft/vscode-cpptools/assets/63085/f98d11a8-7c7d-4d49-9b78-bbc9ec97ef73)

I would be much obliged if `val1` and `val2` kept the color inside the macro definition. That is, the variable.parameter.preprocessor scope should stick to the names inside the macro definition (because these names have that meaning within this macro definition).
